sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -80,22 +81,27 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
     default:
       NewCmd = MGraphBuilder.addCG(std::move(CommandGroup), std::move(Queue));
     }
+    NewEvent = NewCmd->getEvent();
   }
 
   {
-    std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
+    std::unique_lock<std::shared_timed_mutex> Lock(MGraphLock, std::defer_lock);
+    lockSharedTimedMutex(Lock);
```

Comment:
On the other hand, you don't need to unlock and lock again afterwards if you're going to acquire the write lock again.

Reply:
Did you mean that it would be better to have a single lock that will be locked once as it was before?

Reply:
OK, I'll clarify what I mean.
There are two possible approaches:
- employ a existing read-lock + serialization lock;
- employ existing write-lock.

As far as I understood the problem is that 'enqueue command for execution' is potentially modifying operation.

The first approach has an overhead, that's why we would prefer the second one.
Here, at line 83 the write-lock is dropped. Afterwards, write-lock is acquired again at line 89. Though nothing happens in current thread between these lines. Thus I'd prefer to not dropping the write-lock as I don't see any need to 'pause' the 'write-lockness' of the graph.

Also, there are other places where enqueue takes place. We need to change locking there too.

Reply:
I tried to use a single lock in this function and change locking in other places where enqueue takes place, and I see that it leads to failures in some Lit tests. 

The same problem as for commit above "939dc81" - http://ci.llvm.intel.com:8010/#/builders/2/builds/3045
```
terminate called after throwing an instance of 'std::system_error'
  what():  Resource deadlock avoided
```
It requires deep investigation,
So probably, two separate locks are a hotfix.